test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+            this._domain = other._domain;
+            this.data = other.data;
+        }
+
+        operator =(ref lhs: Tensor(?rank,?eltType), rhs: Tensor(rank,eltType)) {
+            lhs._domain = rhs._domain;
+            lhs.data = rhs.data;
+        }
+
+        operator =(ref lhs: Tensor(?rank,?eltType), rhs: [?d] eltType) where d.rank == rank {
+            lhs._domain = d;
+            lhs.data = rhs;
+        }
+
+        proc init=(rhs: [?d] eltType) where d.rank == rank {
+            this.init(d.rank,eltType);
+            this.reshapeDomain(d);
+            this.data = rhs;
+        }
+
+        operator :(from: [?d] ?eltType, type toType: Tensor(d.rank,eltType)) {
+            var t: Tensor(d.rank,eltType) = from;
+            return t;
+        }
+        
+        // ------------ Wasnt sure what to do with these   ----------------
+
+        /*operator =(ref lhs: Tensor(?rank,?eltType), in rhs: ?it) where (isRefIterType(it) || (isArray(rhs) && rhs.eltType == eltType)) && rhs.rank == rank {
+            lhs.reshapeDomain(rhs.domain);
+            lhs.data = rhs;
+        }
+        proc init=(in rhs: ?it) where isRefIterType(it) || isArray(rhs) {
+            this.init(rank,eltType);
+            this.reshapeDomain(rhs.domain);
+            this.data = rhs;
+        }
+        operator :(in from: ?it, type toType: Tensor(?rank,?eltType)) where (isRefIterType(it) || (isArray(from) && from.eltType == eltType)) && from.rank == rank {
+            // compilerError("Cannot convert from ",from.type:string," to ",toType:string);
+            var t: Tensor(rank,eltType) = from;
+            return t;
+        }*/
+
+        operator =(ref lhs: Tensor(?rank,?eltType), itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            lhs.reshapeDomain(itr.domain);
+            lhs.data = itr;
+        }
+
+        proc init=(itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            const A = itr;
+            this.init(A);
+        }
+
+        operator :(itr, type toType: Tensor(?rank,?eltType)) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            var t: Tensor(rank,eltType) = itr;
+            return t;
+        }
+
+
+        // Transposes a vector to a row matrix
+        proc transpose() where rank == 1 {
+            const (p,) = shape;
+            var t = new Tensor(eltType,1,p);
+            t.data[0,..] = this.data;
+            return t;
+        }
+
+        // Matrix transpose
+        proc transpose() where rank == 2 {
+            const (m,n) = this.shape;
+            var M = new Tensor(2,eltType);
+            M.reshapeDomain({0..#n,0..#m});
+            forall (i,j) in M.domain with (ref M, ref this) {
+                M.data[i,j] = this.data[j,i];
+            }
+            return M;
+        }
+
+        // Normalizes the tensor to have unit frobenius norm
+        proc normalize() {
```

Comment:
I'm not seeing where this procedure is used. I could just be missing it, but if it isn't used could you please remove it. Same with any other unused procs in this module that you know of. (Unless they are common enough ML operations that you see a strong argument for including them in the library — in that case it would be nice to have a unit test for them)

Reply:
Yes. The normalization proc specifically is used a lot in other programs, I just haven't been using it.